samcli/commands/build/build_context.py

```diff
@@ -0,0 +1,113 @@
+
+import os
+import shutil
+import yaml
+
+from samcli.yamlhelper import yaml_parse
+from samcli.local.docker.manager import ContainerManager
+from samcli.commands.local.lib.sam_function_provider import SamFunctionProvider
+
+
+class BuildContext(object):
+
+    def __init__(self, template_file,
+                 source_root,
+                 build_dir,
+                 clean=False,
+                 use_container=False,
+                 docker_network=None,
+                 skip_pull_image=False):
+
+        self._template_file = template_file
+        self._source_root = source_root
+        self._build_dir = build_dir
+        self._clean = clean
+        self._use_container = use_container
+        self._docker_network = docker_network
+        self._skip_pull_image = skip_pull_image
+
+        self._function_provider = None
+        self._template_dict = None
+        self._app_builder = None
+        self._container_manager = None
+
+    def __enter__(self):
+        self._template_dict = self._get_template_data(self._template_file)
```

Comment:
Be very cautious here. This template_dict is not run through any transforms or resolves parameters. This could lead to confusion if someone expects one other the other.

Reply:
Agreed, that's why we access only thru providers.